server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -121,28 +207,189 @@ public ClusterState closeIndices(ClusterState currentState, final Index[] indice
+            if (indexRoutingTable == null) {
+                logger.debug("index {} has been blocked before closing but is now deleted, ignoring", index);
+                listener.onResponse(new AcknowledgedResponse(true));
+                return;
+            } else if (indexRoutingTable.allPrimaryShardsUnassigned()) {
+                logger.debug("index {} has been blocked before closing but is now unassigned, ignoring", index);
+                listener.onResponse(new AcknowledgedResponse(true));
+                return;
+            }
+
+            final ImmutableOpenIntMap<IndexShardRoutingTable> shards = indexRoutingTable.getShards();
+            final AtomicArray<AcknowledgedResponse> results = new AtomicArray<>(shards.size());
+            final CountDown countDown = new CountDown(shards.size());
+
+            for (IntObjectCursor<IndexShardRoutingTable> shard : shards) {
+                final IndexShardRoutingTable shardRoutingTable = shard.value;
+                final ShardId shardId = shardRoutingTable.shardId();
+                sendVerifyShardBeforeCloseRequest(shardRoutingTable, timeout, new ActionListener<ReplicationResponse>() {
+                    @Override
+                    public void onResponse(final ReplicationResponse replicationResponse) {
+                        ReplicationResponse.ShardInfo shardInfo = replicationResponse.getShardInfo();
+                        results.setOnce(shardId.id(), new AcknowledgedResponse(shardInfo.getFailed() == 0));
+                        processIfFinished();
+                    }
+
+                    @Override
+                    public void onFailure(final Exception e) {
+                        results.setOnce(shardId.id(), new AcknowledgedResponse(false));
+                        processIfFinished();
+                    }
+
+                    private void processIfFinished() {
+                        if (countDown.countDown()) {
+                            final boolean acknowledged = results.asList().stream().allMatch(AcknowledgedResponse::isAcknowledged);
+                            listener.onResponse(new AcknowledgedResponse(acknowledged));
+                        }
+                    }
+                });
+            }
         }
 
-        //no explicit wait for other nodes needed as we use AckedClusterStateUpdateTask
-        return  allocationService.reroute(
-            ClusterState.builder(updatedState).routingTable(rtBuilder.build()).build(),
-            "indices closed [" + indicesAsString + "]");
+        private void sendVerifyShardBeforeCloseRequest(final IndexShardRoutingTable shardRoutingTable, @Nullable final TimeValue timeout,
+                                                       final ActionListener<ReplicationResponse> listener) {
+            final ShardId shardId = shardRoutingTable.shardId();
+            if (shardRoutingTable.primaryShard().unassigned()) {
```

Comment:
If the primary is unassigned, there's a risk that no shard can be allocated as closed (because the translog might not be flushed). I wonder if we should fail the call to close in that case or whether we should still do the VerifyCloseShardAction, which will then wait for 30 seconds for the shard to be allocated.
@bleskes WDYT?

Reply:
I've done this to keep the current behavior which allows to close indices when shards are not fully allocated. I agree that it can lead to issues if the index is then later reopened as frozen, and in this case the index must be reopened normally, fully allocated and closed again.

I think that this behavior is OK for now. When closed indices will be replicated we will have to change this in order to verify each primary shard using VerifyCloseShardAction and fail the closing if the shard can't be verified.

Reply:
ok, let's add this as a TODO on the meta issue

Reply:
Done